File: internal/controller/solver.go
package controller

import (
	"container/heap"
	"math"
	"sync"

	"github.com/holygeek00/lite-sdwan/pkg/models"
)

// RouteSolver 路径计算引擎
type RouteSolver struct {
	penaltyFactor float64
	hysteresis    float64
	mu            sync.RWMutex
	previousCosts map[string]float64 // "source->target" -> cost
}

// NewRouteSolver 创建新的路径计算引擎
func NewRouteSolver(penaltyFactor, hysteresis float64) *RouteSolver {
	return &RouteSolver{
		penaltyFactor: penaltyFactor,
		hysteresis:    hysteresis,
		previousCosts: make(map[string]float64),
	}
}

// Graph 表示网络拓扑图
type Graph struct {
	nodes map[string]bool
	edges map[string]map[string]float64 // source -> target -> cost
}

// NewGraph 创建新的图
func NewGraph() *Graph {
	return &Graph{
		nodes: make(map[string]bool),
		edges: make(map[string]map[string]float64),
	}
}

// AddNode 添加节点
func (g *Graph) AddNode(id string) {
	g.nodes[id] = true
	if g.edges[id] == nil {
		g.edges[id] = make(map[string]float64)
	}
}

// AddEdge 添加边
func (g *Graph) AddEdge(from, to string, cost float64) {
	g.AddNode(from)
	g.AddNode(to)
	g.edges[from][to] = cost
}

// CalculateCost 计算链路成本
// Cost = RTT_ms + (Loss_rate × PenaltyFactor)
func (s *RouteSolver) CalculateCost(rtt *float64, lossRate float64) float64 {
	if rtt == nil {
		return math.Inf(1) // 链路不可达
	}
	return *rtt + (lossRate * s.penaltyFactor)
}

// BuildGraph 从拓扑数据库构建图
func (s *RouteSolver) BuildGraph(db *TopologyDB) *Graph {
	g := NewGraph()
	allData := db.GetAll()

	// 添加所有节点
	for agentID := range allData {
		g.AddNode(agentID)
	}

	// 添加边
	for source, data := range allData {
		for target, metrics := range data.Metrics {
			cost := s.CalculateCost(metrics.RTT, metrics.Loss)
			g.AddEdge(source, target, cost)
		}
	}

	return g
}

// priorityQueue 用于 Dijkstra 算法的优先队列
type priorityQueue []*pqItem

type pqItem struct {
	node     string
	priority float64
	index    int
}

func (pq priorityQueue) Len() int { return len(pq) }

func (pq priorityQueue) Less(i, j int) bool {
	return pq[i].priority < pq[j].priority
}

func (pq priorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *priorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*pqItem)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *priorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

// DijkstraResult Dijkstra 算法结果
type DijkstraResult struct {
	Distances map[string]float64
	Previous  map[string]string
}

// Dijkstra 执行 Dijkstra 最短路径算法
func (g *Graph) Dijkstra(source string) *DijkstraResult {
	dist := make(map[string]float64)
	prev := make(map[string]string)

	// 初始化距离
	for node := range g.nodes {
		dist[node] = math.Inf(1)
	}
	dist[source] = 0

	// 优先队列
	pq := make(priorityQueue, 0)
	heap.Init(&pq)
	heap.Push(&pq, &pqItem{node: source, priority: 0})

	visited := make(map[string]bool)

	for pq.Len() > 0 {
		item := heap.Pop(&pq).(*pqItem)
		u := item.node

		if visited[u] {
			continue
		}
		visited[u] = true

		// 遍历邻居
		for v, cost := range g.edges[u] {
			if visited[v] {
				continue
			}
			alt := dist[u] + cost
			if alt < dist[v] {
				dist[v] = alt
				prev[v] = u
				heap.Push(&pq, &pqItem{node: v, priority: alt})
			}
		}
	}

	return &DijkstraResult{
		Distances: dist,
		Previous:  prev,
	}
}

// GetPath 从 Dijkstra 结果中获取路径
func (r *DijkstraResult) GetPath(target string) []string {
	if _, ok := r.Previous[target]; !ok && r.Distances[target] == math.Inf(1) {
		return nil // 不可达
	}

	path := []string{target}
	current := target
	for {
		prev, ok := r.Previous[current]
		if !ok {
			break
		}
		path = append([]string{prev}, path...)
		current = prev
	}
	return path
}

// ComputeRoutes 为指定 Agent 计算路由
func (s *RouteSolver) ComputeRoutes(db *TopologyDB, sourceAgent string) []models.RouteConfig {
	g := s.BuildGraph(db)

	// 检查源节点是否存在
	if !g.nodes[sourceAgent] {
		return nil
	}

	result := g.Dijkstra(sourceAgent)
	routes := make([]models.RouteConfig, 0)

	s.mu.Lock()
	defer s.mu.Unlock()

	for target := range g.nodes {
		if target == sourceAgent {
			continue
		}

		path := result.GetPath(target)
		if len(path) < 2 {
			continue // 不可达或就是自己
		}

		newCost := result.Distances[target]
		if math.IsInf(newCost, 1) {
			continue // 不可达
		}

		// 应用迟滞逻辑
		costKey := sourceAgent + "->" + target
		oldCost, exists := s.previousCosts[costKey]

		var nextHop string
		var reason string

		if len(path) == 2 {
			// 直连
			nextHop = "direct"
			reason = "default"
		} else {
			// 需要中继
			nextHop = path[1]
			reason = "optimized_path"
		}

		// 检查是否需要更新路由
		shouldUpdate := false
		if !exists {
			shouldUpdate = true
		} else if newCost < oldCost*(1-s.hysteresis) {
			// 新成本比旧成本低 15% 以上
			shouldUpdate = true
		}

		if shouldUpdate {
			s.previousCosts[costKey] = newCost
			routes = append(routes, models.RouteConfig{
				DstCIDR: target + "/32",
				NextHop: nextHop,
				Reason:  reason,
			})
		}
	}

	return routes
}

// HasLoop 检查路径是否有环
func HasLoop(path []string) bool {
	seen := make(map[string]bool)
	for _, node := range path {
		if seen[node] {
			return true
		}
		seen[node] = true
	}
	return false
}

File: controller/solver.py
"""
Route Solver for Controller.

This module implements the path computation engine that calculates optimal
routes using Dijkstra's algorithm with hysteresis to prevent route flapping.

Requirements: 4.1, 4.2, 4.3, 4.4, 4.5, 7.1, 9.1
"""

from typing import Dict, List, Optional, Tuple
import networkx as nx
from models import RouteConfig


def calculate_cost(latency_ms: float, loss_rate: float, penalty_factor: int = 100) -> float:
    """
    Calculate link cost based on latency and packet loss.
    
    The cost function combines latency and loss rate to produce a single
    metric for path optimization. Higher loss rates are heavily penalized.
    
    Args:
        latency_ms: Round-trip time in milliseconds
        loss_rate: Packet loss rate as decimal (0.0 = no loss, 1.0 = 100% loss)
        penalty_factor: Multiplier for loss rate penalty (default: 100)
    
    Returns:
        Calculated cost value
    
    Formula:
        Cost = Latency_ms + (Loss_rate × Penalty_Factor)
    
    Requirements: 4.1 - Link cost calculation formula
    
    Examples:
        >>> calculate_cost(50.0, 0.0)
        50.0
        >>> calculate_cost(50.0, 0.01)
        51.0
        >>> calculate_cost(100.0, 0.1)
        110.0
    """
    return latency_ms + (loss_rate * penalty_factor)


class RouteSolver:
    """
    Path computation engine for SD-WAN routing.
    
    This class builds a network graph from topology data and computes optimal
    paths using Dijkstra's algorithm. It implements hysteresis to prevent
    route flapping when network conditions fluctuate.
    
    Attributes:
        penalty_factor: Multiplier for packet loss in cost calculation
        hysteresis: Threshold for route switching (0.15 = 15% improvement required)
        previous_costs: Historical path costs for hysteresis logic
    """
    
    def __init__(self, penalty_factor: int = 100, hysteresis: float = 0.15):
        """
        Initialize the route solver.
        
        Args:
            penalty_factor: Multiplier for loss rate in cost calculation (default: 100)
            hysteresis: Minimum improvement threshold for route switching (default: 0.15)
        
        Requirements: 4.1, 4.5, 7.1
        """
        self.penalty_factor = penalty_factor
        self.hysteresis = hysteresis
        self.previous_costs: Dict[Tuple[str, str], float] = {}
    
    def build_graph(self, topology_data: Dict[str, Dict]) -> nx.DiGraph:
        """
        Build a directed graph from topology database data.
        
        Creates a NetworkX DiGraph where nodes represent agents and edges
        represent links with weights calculated from RTT and loss rate.
        
        Args:
            topology_data: Dictionary mapping agent_id to telemetry data
                          Format: {agent_id: {"timestamp": int, "metrics": {...}}}
        
        Returns:
            NetworkX directed graph with weighted edges
        
        Requirements: 4.2 - Graph construction from telemetry
        
        Graph Structure:
            - Nodes: agent IDs (e.g., "10.254.0.1")
            - Edges: directed links with 'weight' attribute (cost)
            - Unreachable links: weight = float('inf')
        """
        G = nx.DiGraph()
        
        # Add all nodes
        for agent_id in topology_data.keys():
            G.add_node(agent_id)
        
        # Add edges with costs
        for source, data in topology_data.items():
            metrics = data.get("metrics", {})
            
            for target, metric_data in metrics.items():
                rtt = metric_data.get("rtt")
                loss = metric_data.get("loss", 0.0)
                
                if rtt is None:  # Link is down
                    cost = float('inf')
                else:
                    cost = calculate_cost(rtt, loss, self.penalty_factor)
                
                G.add_edge(source, target, weight=cost)
        
        return G
    
    def compute_routes_for(self, source_agent: str, topology_data: Dict[str, Dict]) -> List[RouteConfig]:
        """
        Compute optimal routes for a specific agent.
        
        Uses Dijkstra's algorithm to find shortest paths from the source agent
        to all other nodes. Applies hysteresis to prevent route flapping.
        
        Args:
            source_agent: Agent ID requesting routes
            topology_data: Complete topology database data
        
        Returns:
            List of RouteConfig objects specifying next hops
        
        Requirements: 4.3, 4.4, 4.5, 7.1 - Dijkstra algorithm, next hop computation, hysteresis
        
        Route Selection Logic:
            1. Compute shortest path to each destination
            2. Extract next hop (second node in path, or "direct" if adjacent)
            3. Apply hysteresis: only switch if new_cost < old_cost × (1 - threshold)
            4. Generate RouteConfig for each route change
        """
        G = self.build_graph(topology_data)
        routes = []
        
        # Check if source agent exists in graph
        if source_agent not in G:
            return routes
        
        for target in G.nodes():
            if target == source_agent:
                continue
            
            try:
                # Compute shortest path
                path = nx.shortest_path(G, source_agent, target, weight='weight')
                new_cost = nx.shortest_path_length(G, source_agent, target, weight='weight')
                
                # Skip if cost is infinite (unreachable)
                if new_cost == float('inf'):
                    continue
                
                # Apply hysteresis
                cost_key = (source_agent, target)
                old_cost = self.previous_costs.get(cost_key, float('inf'))
                
                # Only switch if new cost is significantly better
                if new_cost < old_cost * (1 - self.hysteresis):
                    # Update cost history
                    self.previous_costs[cost_key] = new_cost
                    
                    # Determine next hop
                    if len(path) == 1:
                        # Source equals target (shouldn't happen, but handle it)
                        continue
                    elif len(path) == 2:
                        # Direct connection
                        next_hop = "direct"
                        reason = "default"
                    else:
                        # Multi-hop path, use second node as next hop
                        next_hop = path[1]
                        reason = "optimized_path"
                    
                    routes.append(RouteConfig(
                        dst_cidr=f"{target}/32",
                        next_hop=next_hop,
                        reason=reason
                    ))
                
            except nx.NetworkXNoPath:
                # No path available, skip this destination
                continue
            except nx.NodeNotFound:
                # Node not in graph, skip
                continue
        
        return routes
    
    def reset_history(self) -> None:
        """
        Reset the cost history.
        
        Useful for testing or when network topology changes significantly.
        """
        self.previous_costs.clear()
    
    def get_cost_history(self) -> Dict[Tuple[str, str], float]:
        """
        Get a copy of the cost history.
        
        Returns:
            Dictionary mapping (source, target) tuples to their last known costs
        """
        return dict(self.previous_costs)

